consensus/src/consensus/inner.rs

```diff
@@ -0,0 +1,508 @@
+    fn verify_transaction(&self, transaction: &SerialTransaction) -> Result<bool, ConsensusError> {
+        if !self
+            .public
+            .parameters
+            .authorized_inner_snark_ids
+            .contains(&to_bytes![transaction.inner_circuit_id]?)
+        {
+            return Ok(false);
+        }
+
+        Ok(InstantiatedDPC::verify(
+            &self.public.public_parameters,
+            &Tx::deserialize(transaction)?,
+            &self.ledger.deserialize::<Components>(),
+        )?)
+    }
+
+    /// Check if the transactions are valid.
+    fn verify_transactions(&self, transactions: &[SerialTransaction]) -> Result<bool, ConsensusError> {
+        for tx in transactions {
+            if !self
+                .public
+                .parameters
+                .authorized_inner_snark_ids
+                .iter()
+                .any(|x| &x[..] == &tx.inner_circuit_id[..])
+            {
+                return Ok(false);
+            }
+        }
+
+        let deserialized = transactions
+            .iter()
+            .map(|t| Tx::deserialize(t))
+            .collect::<IoResult<Vec<_>>>()?;
+
+        Ok(InstantiatedDPC::verify_transactions(
+            &self.public.public_parameters,
+            &deserialized[..],
+            &self.ledger.deserialize::<Components>(),
+        )?)
+    }
+
+    /// Check if the block is valid.
+    /// Verify transactions and transaction fees.
+    async fn verify_block(&self, block: &SerialBlock) -> Result<bool, ConsensusError> {
+        let transaction_ids: Vec<[u8; 32]> = block.transactions.iter().map(|x| x.id).collect();
+        let (merkle_root, pedersen_merkle_root, _) = txids_to_roots(&transaction_ids);
+
+        let canon = self.storage.canon().await?;
+        // Verify the block header
+        if block.header.previous_block_hash != canon.hash {
+            return Err(anyhow!("attempted to commit a block that wasn't a direct child of tip of canon").into());
+        }
+        let parent_header = self.storage.get_block_header(&canon.hash).await?;
+        if let Err(err) =
+            self.public
+                .parameters
+                .verify_header(&block.header, &parent_header, &merkle_root, &pedersen_merkle_root)
+        {
+            error!("block header failed to verify: {:?}", err);
+            return Ok(false);
+        }
+
+        // Verify block amounts and check that there is a single coinbase transaction
+        let mut coinbase_transaction_count: i32 = 0;
+        let mut total_value_balance = AleoAmount::ZERO;
+
+        for transaction in block.transactions.iter() {
+            let value_balance = transaction.value_balance;
+
+            if value_balance.is_negative() {
+                coinbase_transaction_count += 1;
+            }
+
+            total_value_balance = total_value_balance.add(value_balance);
+        }
+        // Check that there is only 1 coinbase transaction
+        if coinbase_transaction_count > 1 {
```

Comment:
This check could be moved into the `if value_balance.is_negative()` above, giving us a quicker potential early return.

Reply:
This was marked as resolved, but it hasn't changed; if having more than one negative `value_balance` is grounds for an early return, we can do it earlier if it's done in the aforementioned check.

Reply:
no we can't, it's in a loop.